web/blog/apache-arrow-as-platform-for-security-data-engineering/index.md

```diff
@@ -15,70 +14,89 @@ Influx's [IOx][iox], DataDog's [Husky][husky], Anyscale's [Ray][ray],
+row-oriented encoding worked well for this.
+
+But as security operations were maturing, requirements extended to analytical
+processing of structured data, making a columnar format increasingly beneficial.
+After having witnessed first-hand the early commitment of [Ray][ray] to Arrow,
+we started using Arrow as optional dependency as additional column-oriented
+encoding. We abstracted a batch of data encoding-independent behind a "table
+slice":
+
+![MsgPack & Arrow](msgpack-arrow.light.png#gh-light-mode-only)
+![MsgPack & Arrow](msgpack-arrow.dark.png#gh-dark-mode-only)
+
+Hiding the concrete encoding behind a cell-based access interface worked for
+low-volume use cases, but backfired as we scaled up and slowed us down
+substantially in development. We needed to make a choice. This is where timing
+was right: our perception of the rapidly evolving Arrow ecosystem changed.
 Arrow-based runtimes were mushrooming all over the place. Nowadays it requires
 only a few lines of code to integrate Arrow data into the central logic of
-applications.
-
-Data interoperability is a sufficient, but not a necessary condition for
-enabling security analytics. We also need to embed the data model of the
-security domain. This is where Arrow's [extension types][extension-types] come
-into play. They add *semantics* to otherwise generic types, e.g., by telling the
-user "this is a transport-layer port" and not just a 16-bit unsigned integer, or
-"this is a connection 4-tuple to represent a network flow" instead of "this is
-a record with 4 fields of type string and unsigned integer". Extension types are
-composable and allow for creating a rich typing layer with meaningful domain
-objects on top of a standardized data representation. Since they are embedded in
-the data, they do not have to be made available out-of-band when crossing the
-boundaries of different tools. Now we have self-describing security data.
+applications. We realized that the primary value proposition of Arrow is to
+*make data interoperability easy*.
+
+But data interoperability is only a sufficient condition for enabling
+sustainable security analytics. The differentiating value of a *security* data
+platform is support for the *security* domain. This is where Arrow's [extension
+types][extension-types] come into play. They add *semantics* to otherwise
+generic types, e.g., by telling the user "this is a transport-layer port" and
+not just a 16-bit unsigned integer, or "this is a connection 4-tuple to
+represent a network flow" instead of "this is a record with 4 fields of type
+string and unsigned integer". Extension types are composable and allow for
+creating a rich typing layer with meaningful domain objects on top of a
+standardized data representation. Since they are embedded in the data, they do
+not have to be made available out-of-band when crossing the boundaries of
+different tools. Now we have self-describing security data.
```

Comment:
This is only half-true, since extension types must be registered so they can be used, and the extra operations can't be pushed down into the arrow library itself. @dominiklohmann correct me if there is some way to extend the API.

Reply:
It sounds like you're talking about "code" whereas I'm talking about "data". Client-libraries with type munching will always be a nicer DX than pure-play Arrow extension types, although you get quite for without it.